snet_cli/commands.py

```diff
@@ -504,3 +506,212 @@ def transact(self):
+        elif not accept_all_defaults:
+            init_args["price"] = input("Choose a price in AGI to call your Service: (default: 0)\n") or init_args["price"]
+
+        if self.args.endpoint:
+            init_args["endpoint"] = self.args.endpoint
+        elif not accept_all_defaults:
+            init_args["endpoint"] = input("URL to call your Service Agent: (default: \"\")\n") or init_args["endpoint"]
+
+        if self.args.tags:
+            init_args["tags"] = self.args.tags
+        elif not accept_all_defaults:
+            init_args["tags"] = input("Input a list of tags for your Service: (default: [])\n") or init_args["tags"]
+
+        if self.args.description:
+            init_args["metadata"]["description"] = self.args.description
+        elif not accept_all_defaults:
+            init_args["metadata"]["description"] = input("Input a description for your Service: (default: none)\n") or init_args["metadata"]["description"]
+
+        with open("service.json", "w") as f:
+            json.dump(init_args, f, indent=4, ensure_ascii=False)
+        self._printout(json.dumps(init_args, indent=4, sort_keys=True))
+        self._printout("\nservice.json file has been created!")
+
+    def publish(self):
+        network = self.w3.version.network
+        if self.args.config:
+            service_json_path = self.args.config
+        else:
+            service_json_path = "service.json"
+
+        with open(service_json_path) as f:
+            service_json = json.load(f)
+
+        # Create model tar and upload it to IPFS
+        if '#' in service_json['model']:
+            model_path = Path("".join(service_json['model'].split('#')[0]))
+        else:
+            model_path = Path(service_json['model'])
+        tmp_f = self._tar_imports(walk_imports(model_path))
+        ipfs_endpoint = urlparse(self.config["ipfs"]["default_ipfs_endpoint"])
+        ipfs_scheme = ipfs_endpoint.scheme if ipfs_endpoint.scheme else "http"
+        ipfs_port = ipfs_endpoint.port if ipfs_endpoint.port else 5001
+        ipfs_client = ipfsapi.connect(urljoin(ipfs_scheme, ipfs_endpoint.hostname), ipfs_port)
+
+        # Upload model to IPFS
+        model_ipfs_hash = ipfs_client.add(read_temp_tar(tmp_f).name)["Hash"]
+        model_ipfs_path = "/ipfs/{}".format(model_ipfs_hash)
+        model_ipfs_uri = uri_reference(model_ipfs_path).copy_with(scheme='ipfs').unsplit()
+
+        # Upload metadata JSON to IPFS with modelURI
+        service_json['metadata']['modelURI'] = model_ipfs_uri
+        with tempfile.NamedTemporaryFile(mode='w+') as tmp_json:
+            json.dump(service_json['metadata'], tmp_json)
+            tmp_json.seek(0)
+            metadata_ipfs_hash = ipfs_client.add(tmp_json.name)["Hash"]
+        metadata_ipfs_path = "/ipfs/{}".format(metadata_ipfs_hash)
+        metadata_ipfs_uri = uri_reference(metadata_ipfs_path).copy_with(scheme='ipfs').unsplit()
+
+        # Create Agent
+        self.args.at = self._getstring("agent_factory_at")
```

Comment:
how would `agent_factory_at` (or `*_at` in any of these commands) be set if it's not declared as an arg on the command? I don't see any `add_contract_identity_arguments` on the new parsers

Reply:
Gets it from session object if it's not in the args
```
    def _getstring(self, name):
        return (self.session.identity.getstring(name) or getattr(self.args, name, None) or
                (self.session.getstring("default_{}".format(name)) or
                 self.session.getstring("current_{}".format(name))))
```